internal/pkg/storage/image.go

```diff
@@ -489,14 +489,21 @@ func (svc *imageService) ResolveNames(systemContext *types.SystemContext, imageN
 		}
 	}
 	// This to prevent any image ID to go through this routine
-	_, err := reference.ParseNormalizedNamed(imageName)
+	ref, err := reference.ParseNormalizedNamed(imageName)
 	if err != nil {
 		if strings.Contains(err.Error(), "cannot specify 64-byte hexadecimal strings") {
 			return nil, ErrCannotParseImageID
 		}
 		return nil, err
 	}
 
+	// Use digest instead of tag when both provided (for docker compatibility)
+	_, isTagged := ref.(reference.NamedTagged)
+	canonical, isDigested := ref.(reference.Canonical)
+	if isTagged && isDigested {
+		imageName = canonical.Name() + "@" + canonical.Digest().String()
```

Comment:
Something like
```go
imageName = reference.WithDigest(reference.TrimNamed(ref), canonical.Digest()).String()
```
please, to use the Go API instead of hard-coding string syntax (compare the other comment).

Reply:
That's better indeed, done.

Reply:
On second thought, oops… neither version is going to work as expected, because both (implied in `.Name()`/`.String()`) convert unqualified repo names (`busybox`) to the fully-qualified form (`docker.io/library/busybox`), breaking the `splitDockerDomain` check immediately below.

Sorry about not noticing that before.

So, it’s either doing this transformation on the _results_ of the search path expansion (a bit more work), or doing it on strings only without using the `reference` subpackage (eww).

(And having non-trivial tests for the unqualified name expansion would certainly have helped.)

Reply:
Oh, right. I improved the tests correspondingly and fixed it.